src/python/pants/backend/shell/shell_command.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from textwrap import dedent
+
+from pants.backend.shell.target_types import (
+    ShellCommandCommandField,
+    ShellCommandLogOutputField,
+    ShellCommandOutputsField,
+    ShellCommandSources,
+    ShellCommandToolsField,
+)
+from pants.core.target_types import FilesSources
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, Directory, MergeDigests, Snapshot
+from pants.engine.process import (
+    BashBinary,
+    BinaryNotFoundError,
+    BinaryPathRequest,
+    BinaryPaths,
+    Process,
+    ProcessResult,
+    SearchPath,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    Sources,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+class GenerateFilesFromShellCommandRequest(GenerateSourcesRequest):
+    input = ShellCommandSources
+    output = FilesSources
+
+
+@rule(desc="Running experimental_shell_command", level=LogLevel.DEBUG)
+async def run_shell_command(
+    request: GenerateFilesFromShellCommandRequest, bash: BashBinary
+) -> GeneratedSources:
+    shell_command = request.protocol_target
+    working_directory = shell_command.address.spec_path
+    command = shell_command[ShellCommandCommandField].value
+    tools = shell_command[ShellCommandToolsField].value
+    outputs = shell_command[ShellCommandOutputsField].value or ()
+
+    if not command:
+        raise ValueError(
+            f"Missing `command` line in `shell_command` target {shell_command.address}."
+        )
+
+    if not tools:
+        raise ValueError(
+            f"Must provide any `tools` used by the `shell_command` {shell_command.address}."
+        )
+
+    tool_requests = [
+        BinaryPathRequest(
+            binary_name=tool,
+            search_path=SearchPath(("/usr/bin", "/bin", "/usr/local/bin")),
+        )
+        for tool in {*tools, *["mkdir", "ln"]}
+    ]
+    tool_paths = await MultiGet(
+        Get(BinaryPaths, BinaryPathRequest, request) for request in tool_requests
+    )
+
+    command_env = {
+        "TOOLS": " ".join(tools),
```

Comment:
It's unlikely, but tools might have spaces in their paths. Should `shlex.escape` them before joining them.